### PR TITLE
Re-introduce "input.enablejoystick" setting

### LINKED ISSUE #105: [Enhancement] xbmc.joystickSuspend()
Title says most of it.
Similar to Kodi's python module functions [xbmc.audioSuspend()](https://codedocs.xyz/xbmc/xbmc/group__python__xbmc.html#ga490c93d077b5fd47106630a92e0e20bb) and [xbmc. audioResume()](https://codedocs.xyz/xbmc/xbmc/group__python__xbmc.html#gadddf5ad858f0e72378115ca2ded3b933)

Rationale:
I can think of a lot of reasons an addon would want to temporarily suspend joystick input (or prevent it all together), but one of the main reasons is to avoid input to Kodi when it's not in focus on the screen (i.e user temporarily brings another application in focus and uses the joystick)

Bonus points if the function can specify by controller type

### PR DESCRIPTION
**Pull request by garbear**
## Description

The `input.enablejoystick` setting was removed in the joystick subsystem introduced in https://github.com/xbmc/xbmc/pull/8807. Instead of this setting, controller input could be disabled from the GUI by disabling the peripheral.joystick add-on.

## Motivation and Context

The reason for re-adding this setting is due to Kodi's interaction with the window manager. On systems without a window manager, Kodi won't receive an event when the main app loses focus. By introducing this setting, an API is provided to clients so they can appropriately enable controller input based on external knowledge.

Discussion: https://forum.kodi.tv/showthread.php?tid=352475

Fixes https://github.com/garbear/xbmc/issues/105

Fixes https://trac.kodi.tv/ticket/16464

## How Has This Been Tested?

Test builds: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200320

(tested on Leia for convenience)

#### Disabling controllers

Ran the command:

```
curl -v --data-binary '{"jsonrpc": "2.0", "method": "Settings.SetSettingValue", "params": {"setting": "input.enablejoystick", "value": false}, "id": 1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc
```

Result in GUI:

Controller disabled

Effect: Controller disabled

#### Enabling controllers

Ran the command:

```
curl -v --data-binary '{"jsonrpc": "2.0", "method": "Settings.SetSettingValue", "params": {"setting": "input.enablejoystick", "value": true}, "id": 1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc
```

Result in GUI:
Controller enabled 2

Effect: Controller enabled

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: xbmc/settings/Settings.cpp [hunk 1]
```diff
@@ -361,6 +361,7 @@ const std::string CSettings::SETTING_AUDIOOUTPUT_VOLUMESTEPS = "audiooutput.volu
 const std::string CSettings::SETTING_INPUT_PERIPHERALS = "input.peripherals";
 const std::string CSettings::SETTING_INPUT_PERIPHERALLIBRARIES = "input.peripherallibraries";
 const std::string CSettings::SETTING_INPUT_ENABLEMOUSE = "input.enablemouse";
+const std::string CSettings::SETTING_INPUT_ENABLE_CONTROLLER = "input.enablejoystick";
```
Comment (lrusak) [inline xbmc/settings/Settings.cpp]: Would it make more sense to make this specific to where it's used rather than generally here?

**Review by lrusak — APPROVED** [review]

**Comment by garbear** [comment]
@lrusak updated

**Comment by garbear** [comment]
Uploaded test builds here: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200320

File: xbmc/input/InputManager.h [hunk 2]
```diff
@@ -289,6 +294,12 @@ class CInputManager : public ISettingCallback,
   std::vector<KODI::MOUSE::IMouseDriverHandler*> m_mouseHandlers;
 
   std::unique_ptr<KODI::KEYBOARD::IKeyboardDriverHandler> m_keyboardEasterEgg;
+
+  // Input state
+  bool m_enableController = true;
+
+  // Settings
+  static const std::string SETTING_INPUT_ENABLE_CONTROLLER;
```
Comment (lrusak) [inline xbmc/input/InputManager.h]: Sorry to bother again. Does this have to be a member if it's only used in this class? I find defining these in an anonymous namespace to be nice and tidy. It's totally up to you though as I'm fine either way :)

**Comment by MilhouseVH** [comment]
@garbear I'll include this in LibreELEC test builds starting with `#0321`

**Comment by phunkyfish** [comment]
Anything to report on this?

**Comment by garbear** [comment]
I'll see what they have to say over in https://github.com/garbear/xbmc/issues/105

**Comment by garbear** [comment]
I fixed the setting being ignored during gameplay. I also rebased on master and fixed the merge conflict.

Finally, I changed "Enable mouse/touch" and "Enable controller" settings to advanced settings. Both of these can kill Kodi input if there is no alternative input method, so we should probably hide them by default.

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

File: xbmc/input/InputManager.cpp [hunk 3]
```diff
@@ -89,6 +92,8 @@ void CInputManager::InitializeInputs()
 
   m_Mouse.Initialize();
   m_Mouse.SetEnabled(CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_INPUT_ENABLEMOUSE));
+
+  m_enableController = CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(SETTING_INPUT_ENABLE_CONTROLLER);
```
Comment (Rechi) [inline xbmc/input/InputManager.cpp:96]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/peripherals/devices/PeripheralJoystick.cpp [hunk 4]
```diff
@@ -203,6 +207,14 @@ bool CPeripheralJoystick::OnButtonMotion(unsigned int buttonIndex, bool bPressed
 
   CSingleLock lock(m_handlerMutex);
 
+  // Check GUI setting and send button release if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
Comment (Rechi) [inline xbmc/peripherals/devices/PeripheralJoystick.cpp:213]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/peripherals/devices/PeripheralJoystick.cpp [hunk 5]
```diff
@@ -246,6 +262,14 @@ bool CPeripheralJoystick::OnHatMotion(unsigned int hatIndex, HAT_STATE state)
 
   CSingleLock lock(m_handlerMutex);
 
+  // Check GUI setting and send hat unpressed if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
Comment (Rechi) [inline xbmc/peripherals/devices/PeripheralJoystick.cpp:268]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/peripherals/devices/PeripheralJoystick.cpp [hunk 6]
```diff
@@ -294,6 +318,14 @@ bool CPeripheralJoystick::OnAxisMotion(unsigned int axisIndex, float position)
 
   CSingleLock lock(m_handlerMutex);
 
+  // Check GUI setting and send analog axis centered if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
Comment (Rechi) [inline xbmc/peripherals/devices/PeripheralJoystick.cpp:324]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by garbear** [comment]
Fixed formatting, but jenkins failed after 10 seconds? jenkins build this please

**Comment by garbear** [comment]
All better. Win32 failure unrelated. Test builds uploaded to: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200516

**Comment by garbear** [comment]
Marked https://trac.kodi.tv/ticket/16464 as solved. Gotta love the 5 year bug fix :)

**Pull request merged**